src/transformers/models/m2m_100/modeling_m2m_100.py

```diff
@@ -1491,3 +1492,284 @@ def _reorder_cache(past_key_values, beam_idx):
+            logits=lm_logits,
+            past_key_values=decoder_outputs.past_key_values,
+            decoder_hidden_states=decoder_outputs.hidden_states,
+            decoder_attentions=decoder_outputs.attentions,
+            cross_attentions=decoder_outputs.cross_attentions,
+            encoder_last_hidden_state=encoder_outputs.last_hidden_state,
+            encoder_hidden_states=encoder_outputs.hidden_states,
+            encoder_attentions=encoder_outputs.attentions,
+        )
+
+    def prepare_inputs_for_generation(
+        self,
+        decoder_input_ids,
+        past_key_values=None,
+        attention_mask=None,
+        head_mask=None,
+        decoder_head_mask=None,
+        cross_attn_head_mask=None,
+        use_cache=None,
+        encoder_outputs=None,
+        **kwargs,
+    ):
+        # cut decoder_input_ids if past is used
+        if past_key_values is not None:
+            past_length = past_key_values[0][0].shape[2]
+
+            # Some generation methods already pass only the last input ID
+            if decoder_input_ids.shape[1] > past_length:
+                remove_prefix_length = past_length
+            else:
+                # Default to old behavior: keep only final ID
+                remove_prefix_length = decoder_input_ids.shape[1] - 1
+
+            decoder_input_ids = decoder_input_ids[:, remove_prefix_length:]
+
+        return {
+            "input_ids": None,  # encoder_outputs is defined. input_ids not needed
+            "encoder_outputs": encoder_outputs,
+            "past_key_values": past_key_values,
+            "decoder_input_ids": decoder_input_ids,
+            "attention_mask": attention_mask,
+            "head_mask": head_mask,
+            "decoder_head_mask": decoder_head_mask,
+            "cross_attn_head_mask": cross_attn_head_mask,
+            "use_cache": use_cache,  # change this to avoid caching (presumably for debugging)
+        }
+
+    @staticmethod
+    def _reorder_cache(past_key_values, beam_idx):
+        reordered_past = ()
+        for layer_past in past_key_values:
+            reordered_past += (
+                tuple(past_state.index_select(0, beam_idx.to(past_state.device)) for past_state in layer_past),
+            )
+        return reordered_past
+
+
+@add_start_docstrings(
+    "The M2M100 (SONAR) transformer model outputting encoder's raw hidden states. Can be used for embedding sentences.",
+    M2M_100_START_DOCSTRING,
+)
+class M2M100EncoderModel(M2M100PreTrainedModel):
+    _tied_weights_keys = ["encoder.embed_tokens.weight"]
+    _keys_to_ignore_on_load_unexpected = [r"decoder"]
+
+    def __init__(self, config: M2M100Config):
+        super().__init__(config)
+        self.shared = nn.Embedding(config.vocab_size, config.d_model)
+
+        encoder_config = copy.deepcopy(config)
+        encoder_config.use_cache = False
+        encoder_config.is_encoder_decoder = False
+        self.encoder = M2M100Encoder(encoder_config, self.shared)
```

Comment:
this is a bit weird. If you pass the shared nn.emebdding, only the weights are used, and it's otherwise a sinusoidal : https://github.com/huggingface/transformers/blob/6cffc90dcacca489f785fa4689addbc49f8dd17f/src/transformers/models/m2m_100/modeling_m2m_100.py#L796-L797 which begs the question to the need for this whole class! ?

Reply:
This whole `M2M100EncoderModel` class has several rationales to exist:
1. The additional code in `forward` for pooling the token embeddings into a sentence embedding.
2. Having the same level of nesting the parameters (model->encoder->its layers) makes it easier to convert a trained encoder-decoder model (having the architecture of M2M100/NLLB) into a pair of separated encoder and decoder: one needs only to select the required parameters from the state dict, without having to rename them.
3. We can put an appropriate docstring on it :-)

Reply:
Okay, IMO only 1. is a strong reason, I don't know how much `pool_last_hidden_state` are used but okay. 
2. is all in the conversion script no?

Reply:
1. Applying `pool_last_hidden_state` is going to be the main use case: producing pooled sentence embeddings. I am setting this parameter by default to `False` only for consistency with other encoders.